### PR TITLE
fix memory leak with a descriptor

### PR DESCRIPTION
**Pull request by frosty00**
I think we should define the descriptor in another file since it is pretty ugly in the same file as the base exchange... 

let's see if it passes the test ;0

### CODE REVIEW

**Comment by kroitor** [comment]
@frosty00 btw, have you checked the /async_support/ subfolder for the same bug?

**Comment by frosty00** [comment]
just checked now, it seems my fix fixed that file also (because it inherits from the synchronous exchange), not sure exactly how, but there is less leaking memory.

**Comment by frosty00** [comment]
I'm using @dingyaguang117 's script to test it:


```
import gc
import time
import os
import sys
import psutil
import ccxt

def k():
    gc.enable()
    gc.set_debug(gc.DEBUG_COLLECTABLE | gc.DEBUG_UNCOLLECTABLE | gc.DEBUG_SAVEALL)

    # MethoTest()
    loop = 10
    while loop > 0:
        exchange = ccxt.async_support.poloniex({
            'enableRateLimit': True,
        })
        process = psutil.Process(os.getpid())
        print(exchange.iso8601(exchange.milliseconds()),
              process.memory_info()[0])
        del exchange  # --------- does not really call __del__ here
        loop -= 1

    print('gc-ing...')
    _unreachable = gc.collect()
    print('unreachable num: %d' % _unreachable)
    print('leak num: %d' % len(gc.garbage))

    time.sleep(3)
```

**Comment by frosty00** [comment]
this has a bug (doesn't work with multiple instances), will fix tomorrow

**Comment by frosty00** [comment]
This branch still has bugs (e.g. inheritance since the rest api isn't defined) but I'll work on it

**Comment by kroitor** [comment]
> However, upon reading an attribute such as exchange.public_get_markets the partial needs to be initiated each time.

Ouch... hope we can find a workaround for this...

**Comment by frosty00** [comment]
So I found a workaround... What is going on here is actually the descriptor protocol behind the scenes. But instead of using partials I used a closure to capture the arguments in place - partials do not have a `__get__` method. Methods are just functions that are stored in the a class's `__dict__`. Functions, by default, in python implement the descriptor protocol. `instance.method_call(*args)` is converted to `type(instance).__dict__['method_call'].__get__(instance, type(instance))(*args)`

If a function is inside the `__dict__` of a class it is a method, `klass.method` will return a function and `instance.method` will return the bounded method, via the descriptor protocol.

**Comment by ngugcx** [comment]
@frosty00 ，have you tested with multiple exchanges? Just a guess, is it possible that functions may be overwritten in some cases and the class will be inflating with many exchanges?

File: python/ccxt/base/exchange.py [hunk 1]
```diff
@@ -361,10 +348,20 @@ def define_rest_api(cls, api, method_name, options={}):
...
+                            if len(args) > 1:
+                                inner_kwargs['params'] = args[1]
+                                args = [args[0]]  # a reference to self without circular dependencies
+                            return entry(*args, **inner_kwargs)
+                        return inner
+                    to_bind = partialer()
+                    setattr(cls, camelcase, to_bind)
+                    setattr(cls, underscore, to_bind)
         cls.rest_api_defined = True
         # ^ this code only runs once per class
```
Comment (ngugcx) [inline python/ccxt/base/exchange.py]: Will lose the capability to add rest api at runtime?

Comment (frosty00) [reply python/ccxt/base/exchange.py]: You can still modify the rest api at runtime, but only by decorating (wrapping with another function) `.describe` before initiating any instances. 

```
exchange_class = ccxt.bittrex()
exchange_class.describe = decorator(exchange_class.describe)
exchange_class()  # no more rest api's can be defined after this
```

but I think you'd be better off using the descriptors or binding your own custom rest api functions than trying to do it like this

**Comment by frosty00** [comment]
just ran my script and it works:

**Comment by frosty00** [comment]
@kroitor can we merge this

just tested this branch on my script and it works as usual https://i.imgur.com/0mRh0ox.png :]

**Comment by kroitor** [comment]
> @kroitor can we merge this

I don't really like this: https://github.com/ccxt/ccxt/pull/4394#discussion_r244479303

**Comment by frosty00** [comment]
yeah I changed it in https://github.com/ccxt/ccxt/pull/4394/commits/6af239a1e12a0ffccc1bc17c1a235e8283a9c16e

look at the most recent changes to the code

also working on a ccxt website will release today or tomorrow

**Comment by frosty00** [comment]
Another reason to merge this is that the parameter info (⌘P) in jetbrains IDEs for generated methods is now accurate, before it just said *args, **kwargs

**Pull request merged**